src/storage/src/render/sources.rs

```diff
@@ -70,12 +77,44 @@ where
     G: Scope<Timestamp = Timestamp>,
 {
     let ((handle, capability), ok_stream) = scope.new_unordered_input::<(Row, Timestamp, Diff)>();
+    // Convert to reference counted, so that users can downgrade it and allow the
+    // following code to destroy it later on.
+    let capability = Rc::new(RefCell::new(capability));
     let err_collection = Collection::empty(scope);
 
     let as_of_frontier = as_of_frontier.clone();
+    let mut vector = Vec::new();
+    let mut token = None;
+
+    // Note: this `unary` operator is adapted from the source for `Operator::map_in_place`.
+    // It adds a way to drop the operator's core capability, if the returned, thread-safe
+    // `SourceToken` is dropped.
     let ok_collection = ok_stream
-        .map_in_place(move |(_, time, _)| {
-            time.advance_by(as_of_frontier.borrow());
+        .unary(Pipeline, "MapInPlaceWithSyncToken", |_, operator_info| {
```

Comment:
Why is this suddenly complicated?

Reply:
I need access to the underlying `operator_info` to make a `SyncActivator` for the operator, and to add the `upgrade` check

Reply:
@frankmcsherry If I leave it, where would I put the `upgrade().is_none() -> drop cap` logic?

Reply:
I read this some more, and I think the answer I was looking for was: "The `new_unordered_input` operator does not give us the opportunity to drop its capability; we need an operator that can drop this capability, and rather than introduce a new one I fused it with a custom implementation of `map_in_place`". Imo this could be a fair bit clearer with the capability dropping operator right after the input is constructed, leaving `map_in_place` untouched (especially given that we may end up ripping this all out in a week, once we can use `Rc` again).

Reply:
In particular, this text
> It adds a way to drop the operator's core capability, 
led me to believe it was *this* operator's capability, when it is instead `capability` from the unordered input that we want to drop, afaict.

Reply:
@frankmcsherry Ill add a comment here explaining this better

I agree it might be cleaner with a "cap dropping operator" that is before the `map_in_place`, but will leave it because, as you said, this may be ripped up relatively soon!

Reply:
Well, we can rip it out, we'd have to replace it with the prior `map_in_place` operator. Having the separate operator would allow that to be deleted without needing to reconstruct the prior state.

Reply:
good point, I updated this pr to do just that! wasn't too bad!